File: services/notification/src/observability.ts
/**
 * Observabilidad del Servicio de Notificación (§Observabilidad, research D-12).
 *
 * Expone tres cosas en un puerto HTTP propio:
 *
 * - `/healthz` — ¿el proceso está vivo? Kubernetes reinicia el pod si falla.
 * - `/readyz`  — ¿puede trabajar? Kubernetes le quita tráfico si falla.
 * - `/metrics` — latencia, tasa de error y throughput en formato Prometheus.
 *
 * La distinción entre las dos primeras sondas es la que evita el peor fallo operativo
 * de todos: si `/healthz` comprobara la base de datos, una caída de PostgreSQL
 * reiniciaría TODAS las réplicas a la vez, y al volver la base se encontraría con
 * ocho procesos arrancando en frío en lugar de ocho listos para reanudar.
 *
 * Que este servicio abra un puerto HTTP no contradice el Principio II. No hay
 * superficie REST de negocio: son sondas de infraestructura, en un puerto distinto del
 * de servicio, y no exponen ningún dato de dominio.
 *
 * Las métricas se escriben a mano en lugar de traer un cliente de Prometheus: lo que
 * hace falta son un contador y un histograma, y una librería con registries,
 * colectores y push-gateways sería más configuración que la que sustituye.
 */
import { createServer, type Server } from 'node:http';

import type { Logger } from './logger.js';

/**
 * Cotas del histograma de latencia, en segundos.
 *
 * Están elegidas alrededor de SC-007 (95 % de las notificaciones entregadas en menos
 * de dos minutos): sin una cota cerca del objetivo, el percentil que hay que vigilar
 * caería siempre dentro del mismo intervalo y la métrica no distinguiría cumplir de
 * incumplir.
 */
const LATENCY_BUCKETS = [0.1, 0.5, 1, 2, 5, 10, 30, 60, 120];

/**
 * Sumidero de métricas.
 *
 * Es una interfaz para que el despachador no dependa del registro concreto: sus
 * pruebas no tienen por qué acumular histogramas para comprobar tres transiciones.
 */
export interface MetricsSink {
  observe(operation: string, outcome: 'success' | 'failure', seconds: number): void;
}

/** Registro de métricas del proceso. */
export class Metrics implements MetricsSink {
  private readonly counts = new Map<string, number>();
  private readonly latencies = new Map<string, number[]>();

  /**
   * Registra el desenlace de una operación y su duración.
   *
   * Throughput y tasa de error salen del mismo contador porque son la misma cuenta
   * partida por `outcome`: registrar dos contadores independientes permitiría que
   * discreparan.
   */
  public observe(operation: string, outcome: 'success' | 'failure', seconds: number): void {
    const key = `${operation}|${outcome}`;
    this.counts.set(key, (this.counts.get(key) ?? 0) + 1);

    const samples = this.latencies.get(operation) ?? [];
    samples.push(seconds);
    this.latencies.set(operation, samples);
  }

  /** Serializa el registro en el formato de texto de Prometheus. */
  public render(): string {
    const lines: string[] = [
      '# HELP notification_operations_total Operaciones procesadas por desenlace.',
      '# TYPE notification_operations_total counter',
    ];
    for (const [key, value] of this.counts) {
      const [operation, outcome] = key.split('|');
      lines.push(
        `notification_operations_total{operation="${operation}",outcome="${outcome}"} ${value}`,
      );
    }

    lines.push(
      '# HELP notification_operation_duration_seconds Duración de cada operación.',
      '# TYPE notification_operation_duration_seconds histogram',
    );
    for (const [operation, samples] of this.latencies) {
      let cumulative = 0;
      for (const bound of LATENCY_BUCKETS) {
        cumulative = samples.filter((s) => s <= bound).length;
        lines.push(
          `notification_operation_duration_seconds_bucket{operation="${operation}",le="${bound}"} ${cumulative}`,
        );
      }
      lines.push(
        `notification_operation_duration_seconds_bucket{operation="${operation}",le="+Inf"} ${samples.length}`,
        `notification_operation_duration_seconds_sum{operation="${operation}"} ${samples.reduce((a, b) => a + b, 0)}`,
        `notification_operation_duration_seconds_count{operation="${operation}"} ${samples.length}`,
      );
    }

    return `${lines.join('\n')}\n`;
  }
}

/** Qué debe estar en pie para considerar el proceso listo. */
export interface ReadinessCheck {
  (): Promise<boolean>;
}

/**
 * Arranca el servidor de sondas y métricas.
 *
 * @returns el servidor, para cerrarlo en el apagado ordenado.
 */
export function startProbeServer(
  port: number,
  metrics: Metrics,
  isReady: ReadinessCheck,
  logger: Logger,
): Server {
  const server = createServer((req, res) => {
    const path = (req.url ?? '/').split('?')[0];

    if (path === '/healthz') {
      // Vivacidad: NO consulta dependencias. Si lo hiciera, una caída de PostgreSQL
      // reiniciaría todas las réplicas a la vez.
      res.writeHead(200, { 'Content-Type': 'text/plain' }).end('ok\n');
      return;
    }

    if (path === '/metrics') {
      res.writeHead(200, { 'Content-Type': 'text/plain; version=0.0.4' }).end(metrics.render());
      return;
    }

    if (path === '/readyz') {
      void isReady()
        .then((ready) => {
          res
            .writeHead(ready ? 200 : 503, { 'Content-Type': 'text/plain' })
            .end(ready ? 'ready\n' : 'not ready\n');
        })
        .catch((err: unknown) => {
          logger.error('la sonda de readiness falló', {
            error: err instanceof Error ? err.message : String(err),
          });
          res.writeHead(503, { 'Content-Type': 'text/plain' }).end('not ready\n');
        });
      return;
    }

    res.writeHead(404, { 'Content-Type': 'text/plain' }).end('not found\n');
  });

  server.listen(port, () => logger.info('sondas de salud escuchando', { port }));
  return server;
}

File: services/simulator/src/observability.rs
//! Observabilidad del Simulador (§Observabilidad, research D-12).
//!
//! Expone en un puerto HTTP propio las tres cosas que la constitución exige de todos
//! los servicios y que aquí faltaban —el log JSON ya lo monta `main.rs`:
//!
//! - `/healthz` — ¿el proceso está vivo? Kubernetes REINICIA el pod si falla.
//! - `/readyz`  — ¿puede trabajar? Kubernetes le QUITA TRÁFICO si falla.
//! - `/metrics` — latencia, tasa de error y throughput en formato Prometheus.
//!
//! La distinción entre las dos sondas evita el peor fallo operativo posible: si
//! `/healthz` comprobara PostgreSQL, una caída de la base reiniciaría TODAS las
//! réplicas a la vez y, al volver, se encontraría con procesos arrancando en frío en
//! lugar de con réplicas listas para atender.
//!
//! Que un servicio gRPC abra un puerto HTTP no contradice el Principio II: no hay
//! superficie REST de negocio, son sondas de infraestructura en un puerto aparte.
//!
//! ## Por qué el servidor HTTP está escrito a mano
//!
//! Un `hyper` completo aquí traería un árbol de dependencias entero para servir tres
//! respuestas de texto sin cabeceras ni negociación de contenido. Lo que hace falta es
//! leer una línea de petición y contestar, y eso cabe en este archivo.
//!
//! ## Por qué las latencias son enteros
//!
//! `f64` está PROHIBIDO en este crate (`clippy::disallowed_types`, Principio VIII). La
//! prohibición apunta a dinero y tasas, pero en lugar de pedir una excepción para los
//! segundos se acumulan MICROSEGUNDOS en `u64` y se formatean al renderizar. El
//! resultado es idéntico para Prometheus y el crate se queda sin una sola excepción a
//! la regla, que es más fácil de defender que una excepción bien argumentada.

use std::collections::HashMap;
use std::future::Future;
use std::sync::{Mutex, OnceLock};
use std::time::Duration;

use sqlx::PgPool;
use tokio::io::{AsyncBufReadExt, AsyncWriteExt, BufReader};
use tokio::net::TcpListener;
use tracing::{error, info, warn};

/// Nombre del servicio en las métricas.
const SERVICE: &str = "simulator";

/// Puerto de las sondas cuando `HEALTH_PORT` no está.
///
/// Tiene valor por defecto —al contrario que `GRPC_PORT`, que es obligatorio— porque
/// un despliegue que lo olvide debe quedarse sin sondas, no sin arrancar.
pub const DEFAULT_HEALTH_PORT: &str = "8080";

/// Cotas del histograma, en microsegundos, con su etiqueta en segundos.
///
/// Están elegidas alrededor de SC-006 (una simulación responde en menos de 500 ms):
/// sin una cota cerca del umbral que se vigila, el percentil caería siempre dentro del
/// mismo intervalo y la métrica no distinguiría cumplirlo de incumplirlo.
const BUCKETS: &[(u64, &str)] = &[
    (5_000, "0.005"),
    (10_000, "0.01"),
    (25_000, "0.025"),
    (50_000, "0.05"),
    (100_000, "0.1"),
    (250_000, "0.25"),
    (500_000, "0.5"),
    (1_000_000, "1"),
    (2_500_000, "2.5"),
    (5_000_000, "5"),
    (10_000_000, "10"),
];

/// Estado acumulado de las métricas.
#[derive(Default)]
struct Registry {
    /// (operación, código) → número de peticiones.
    ///
    /// Throughput y tasa de error salen de este mismo mapa porque son la misma cuenta
    /// partida por el código: dos contadores independientes podrían discrepar.
    counts: HashMap<(String, String), u64>,
    /// operación → duraciones en microsegundos.
    samples: HashMap<String, Vec<u64>>,
}

fn registry() -> &'static Mutex<Registry> {
    static REGISTRY: OnceLock<Mutex<Registry>> = OnceLock::new();
    REGISTRY.get_or_init(|| Mutex::new(Registry::default()))
}

/// Registra el desenlace y la duración de una operación.
pub fn observe(operation: &str, code: &str, elapsed: Duration) {
    let micros = u64::try_from(elapsed.as_micros()).unwrap_or(u64::MAX);

    // `unwrap_or_else` sobre el veneno del mutex: si otro hilo entró en pánico
    // sosteniéndolo, seguir contando métricas es preferible a propagar el pánico y
    // tumbar el servicio por su instrumentación.
    let mut reg = registry()
        .lock()
        .unwrap_or_else(|poisoned| poisoned.into_inner());
    *reg.counts
        .entry((operation.to_owned(), code.to_owned()))
        .or_insert(0) += 1;
    reg.samples
        .entry(operation.to_owned())
        .or_default()
        .push(micros);
}

/// Serializa el registro en el formato de texto de Prometheus.
fn render() -> String {
    let reg = registry()
        .lock()
        .unwrap_or_else(|poisoned| poisoned.into_inner());
    let mut out = String::new();

    out.push_str("# HELP fintcart_requests_total Peticiones atendidas, por operación y código.\n");
    out.push_str("# TYPE fintcart_requests_total counter\n");
    for ((operation, code), value) in &reg.counts {
        out.push_str(&format!(
            "fintcart_requests_total{{service=\"{SERVICE}\",operation=\"{operation}\",code=\"{code}\"}} {value}\n"
        ));
    }

    out.push_str("# HELP fintcart_request_duration_seconds Latencia de las operaciones.\n");
    out.push_str("# TYPE fintcart_request_duration_seconds histogram\n");
    for (operation, samples) in &reg.samples {
        for (bound_micros, label) in BUCKETS {
            let count = samples.iter().filter(|s| *s <= bound_micros).count();
            out.push_str(&format!(
                "fintcart_request_duration_seconds_bucket{{service=\"{SERVICE}\",operation=\"{operation}\",le=\"{label}\"}} {count}\n"
            ));
        }
        let total: u64 = samples.iter().sum();
        out.push_str(&format!(
            "fintcart_request_duration_seconds_bucket{{service=\"{SERVICE}\",operation=\"{operation}\",le=\"+Inf\"}} {}\n\
             fintcart_request_duration_seconds_sum{{service=\"{SERVICE}\",operation=\"{operation}\"}} {}\n\
             fintcart_request_duration_seconds_count{{service=\"{SERVICE}\",operation=\"{operation}\"}} {}\n",
            samples.len(),
            seconds(total),
            samples.len(),
        ));
    }

    out
}

/// Formatea microsegundos como segundos con seis decimales, sin punto flotante.
fn seconds(micros: u64) -> String {
    format!("{}.{:06}", micros / 1_000_000, micros % 1_000_000)
}

/// Sirve las sondas hasta que `shutdown` resuelva.
///
/// Un fallo al abrir el puerto NO tumba el proceso: sin `/readyz`, Kubernetes deja de
/// mandarle tráfico, que es la degradación deseada. Matar el servicio porque su puerto
/// de DIAGNÓSTICO no arrancó invertiría la relación entre el servicio y lo que lo
/// observa.
pub async fn serve_probes(port: &str, pool: PgPool, shutdown: impl Future<Output = ()> + Send) {
    let addr = format!("0.0.0.0:{port}");
    let listener = match TcpListener::bind(&addr).await {
        Ok(listener) => listener,
        Err(err) => {
            error!(%addr, error = %err, "no se pudieron abrir las sondas de salud");
            return;
        }
    };
    info!(%addr, "sondas de salud escuchando");

    tokio::pin!(shutdown);
    loop {
        tokio::select! {
            () = &mut shutdown => {
                info!("sondas de salud detenidas");
                return;
            }
            accepted = listener.accept() => {
                match accepted {
                    Ok((stream, _)) => {
                        let pool = pool.clone();
                        // Una conexión por tarea: una sonda lenta —la de readiness
                        // consulta la base— no puede bloquear a las demás.
                        tokio::spawn(async move { handle(stream, pool).await; });
                    }
                    Err(err) => warn!(error = %err, "conexión de sonda rechazada"),
                }
            }
        }
    }
}

/// Atiende una petición de sonda.
async fn handle(stream: tokio::net::TcpStream, pool: PgPool) {
    let mut reader = BufReader::new(stream);
    let mut request_line = String::new();

    if reader.read_line(&mut request_line).await.is_err() {
        return;
    }

    // `GET /readyz HTTP/1.1` → `/readyz`. Basta con la ruta: estas sondas no negocian
    // contenido ni leen cabeceras, y pretender lo contrario sería escribir un servidor
    // HTTP de verdad para servir tres cadenas.
    let path = request_line.split_whitespace().nth(1).unwrap_or("/");

    let (status, body) = match path {
        // Vivacidad: NO consulta ninguna dependencia. Ver la cabecera del módulo.
        "/healthz" => ("200 OK", "ok\n".to_owned()),
        "/metrics" => ("200 OK", render()),
        "/readyz" => match ready(&pool).await {
            Ok(()) => ("200 OK", "ready\n".to_owned()),
            Err(err) => {
                warn!(error = %err, "sonda de readiness negativa");
                ("503 Service Unavailable", "not ready\n".to_owned())
            }
        },
        _ => ("404 Not Found", "not found\n".to_owned()),
    };

    // `Connection: close` evita tener que implementar keep-alive: el cliente de estas
    // sondas es kubelet o Prometheus, y ambos abren una conexión por comprobación.
    let response = format!(
        "HTTP/1.1 {status}\r\nContent-Type: text/plain; charset=utf-8\r\nContent-Length: {}\r\nConnection: close\r\n\r\n{body}",
        body.len()
    );
    let _ = reader.get_mut().write_all(response.as_bytes()).await;
    let _ = reader.get_mut().shutdown().await;
}

/// Plazo de la comprobación de readiness.
///
/// Sin plazo, una base que acepta la conexión y no responde dejaría la sonda colgada;
/// kubelet la daría por fallida de todos modos, pero varios segundos más tarde y sin
/// nada en el log que lo explicara.
const READY_TIMEOUT: Duration = Duration::from_secs(3);

/// Comprueba que el servicio puede trabajar.
async fn ready(pool: &PgPool) -> anyhow::Result<()> {
    let probe = sqlx::query("SELECT 1").execute(pool);
    match tokio::time::timeout(READY_TIMEOUT, probe).await {
        Ok(Ok(_)) => Ok(()),
        Ok(Err(err)) => Err(anyhow::anyhow!("simulator_db no responde: {err}")),
        Err(_) => Err(anyhow::anyhow!(
            "simulator_db no respondió en {READY_TIMEOUT:?}"
        )),
    }
}
